packages/core/crates/core-types/src/channels.rs

```diff
@@ -204,195 +214,369 @@ impl Default for Ticket {
+        self.signature = Some(Signature::sign_message(
+            &self.get_hash(domain_separator).to_bytes(),
+            signing_key,
+        ));
     }
 
     /// Convenience method for creating a zero-hop ticket
-    pub fn new_zero_hop(destination: Address, private_key: &ChainKeypair) -> Self {
+    pub fn new_zero_hop(destination: Address, private_key: &ChainKeypair, domain_separator: &Hash) -> Self {
         Self::new(
+            private_key.public().to_address(),
             destination,
-            U256::zero(),
             Balance::new(0u32.into(), BalanceType::HOPR),
             U256::zero(),
             U256::zero(),
+            0.0,
+            U256::zero(),
+            EthereumChallenge::default(),
             private_key,
+            domain_separator,
         )
-    }
-
-    /// Serializes the ticket except the signature
-    pub fn serialize_unsigned(&self) -> Box<[u8]> {
-        Self::serialize_unsigned_aux(
-            &self.counterparty,
-            &self.challenge,
-            &self.amount,
-            &self.win_prob,
-            &self.index,
-            &self.channel_epoch,
-        )
-        .into_boxed_slice()
-    }
-
-    /// Computes Ethereum signature hash of the ticket
-    pub fn get_hash(&self) -> Hash {
-        ethereum_signed_hash(Hash::create(&[&self.serialize_unsigned()]).to_bytes())
-    }
-
-    /// Computes a candidate check value to verify if this ticket is winning
-    pub fn get_luck(&self, preimage: &Hash, channel_response: &Response) -> U256 {
-        U256::from_bytes(
-            &Hash::create(&[
-                &self.get_hash().to_bytes(),
-                &preimage.to_bytes(),
-                &channel_response.to_bytes(),
-            ])
-            .to_bytes(),
-        )
-        .unwrap()
-    }
-
-    /// Decides whether a ticket is a win or not.
-    /// Note that this mimics the on-chain logic.
-    /// Purpose of the function is to check the validity of ticket before we submit it to the blockchain.
-    pub fn is_winning(&self, preimage: &Hash, channel_response: &Response, win_prob: U256) -> bool {
-        let luck = self.get_luck(preimage, channel_response);
-        luck.value().le(win_prob.value())
+        .expect("Failed to create zero-hop ticket")
     }
 
     /// Based on the price of this ticket, determines the path position (hop number) this ticket
     /// relates to.
-    pub fn get_path_position(&self, price_per_packet: U256, inverse_ticket_win_prob: U256) -> u8 {
-        let base_unit = price_per_packet.value().mul(inverse_ticket_win_prob.value());
-        self.amount.value().value().div(base_unit).as_u8()
+    pub fn get_path_position(&self, price_per_packet: U256, win_prob: f64) -> u8 {
+        let base_unit = if win_prob == 1.0 {
+            price_per_packet
+        } else {
+            // win_prob between 0.0 and 1.0
+            let bitmask = hex!("000fffffffffffff"); // clear IEEE754 sign bit and exponent
```

Comment:
Why not assigning the value to u64 directly? e.g. `let bitmask = 0xabcdef_u64;`

Reply:
only works if the bitstring fully fits into native types, but good point here